File: base/Sources/ImportAsset.ts

class ImportAsset {

	static run = (path: string, dropX: f32 = -1.0, dropY: f32 = -1.0, showBox: bool = true, hdrAsEnvmap: bool = true, done: ()=>void = null) => {

		if (path.startsWith("cloud")) {
			let do_cache_cloud = () => {
				file_cache_cloud(path, (abs: string) => {
					if (abs == null) return;
					ImportAsset.run(abs, dropX, dropY, showBox, hdrAsEnvmap, done);
				});
			}

			///if (krom_android || krom_ios)
			base_notify_on_next_frame(() => {
				console_toast(tr("Downloading"));
				base_notify_on_next_frame(do_cache_cloud);
			});
			///else
			do_cache_cloud();
			///end

			return;
		}

		if (path_is_mesh(path)) {
			showBox ? project_import_mesh_box(path) : ImportMesh.run(path);
			if (dropX > 0) UIBox.click_to_hide = false; // Prevent closing when going back to window after drag and drop
		}
		else if (path_is_texture(path)) {
			ImportTexture.run(path, hdrAsEnvmap);
			// Place image node
			let x0: i32 = UINodes.wx;
			let x1: i32 = UINodes.wx + UINodes.ww;
			if (UINodes.show && dropX > x0 && dropX < x1) {
				let asset_index: i32 = 0;
				for (let i: i32 = 0; i < project_assets.length; ++i) {
					if (project_assets[i].file == path) {
						asset_index = i;
						break;
					}
				}
				UINodes.accept_asset_drag(asset_index);
				UINodes.get_nodes().nodesDrag = false;
				UINodes.hwnd.redraws = 2;
			}

			///if is_paint
			if (context_raw.tool == workspace_tool_t.COLORID && project_asset_names.length == 1) {
				UIHeader.header_handle.redraws = 2;
				context_raw.ddirty = 2;
			}
			///end
		}
		else if (path_is_project(path)) {
			ImportArm.run_project(path);
		}
		else if (path_is_plugin(path)) {
			ImportPlugin.run(path);
		}
		else if (path_is_gimp_color_palette(path)) {
			ImportGpl.run(path, false);
		}
		///if is_paint
		else if (path_is_font(path)) {
			ImportFont.run(path);
		}
		else if (path_is_folder(path)) {
			ImportFolder.run(path);
		}
		///end
		else {
			if (context_enable_import_plugin(path)) {
				ImportAsset.run(path, dropX, dropY, showBox);
			}
			else {
				console_error(strings_error1());
			}
		}

		if (done != null) done();
	}
}

File: paint/sources/import_asset.c

#include "global.h"

void import_asset_run_cache_cloud_done(char *abs) {
	if (abs == NULL) {
		return;
	}
	import_asset_run(abs, _import_asset_drop_x, _import_asset_drop_y, _import_asset_show_box, _import_asset_hdr_as_envmap, _import_asset_done);
}

void import_asset_run(char *path, f32 drop_x, f32 drop_y, bool show_box, bool hdr_as_envmap, void (*done)(void)) {
	if (starts_with(path, "cloud")) {
		#ifdef IRON_ANDROID
		console_toast(tr("Downloading"));
		#else
		console_info(tr("Downloading"));
		#endif

		_import_asset_drop_x        = drop_x;
		_import_asset_drop_y        = drop_y;
		_import_asset_show_box      = show_box;
		_import_asset_hdr_as_envmap = hdr_as_envmap;
		gc_unroot(_import_asset_done);
		_import_asset_done = done;
		gc_root(_import_asset_done);
		file_cache_cloud(path, &import_asset_run_cache_cloud_done, config_raw->server);

		return;
	}

	if (path_is_mesh(path)) {
		if (context_raw->tool == TOOL_TYPE_GIZMO) {
			// project_import_mesh_box(path, false, false, tab_meshes_import_mesh_done);
			project_import_mesh_box(path, false, false, tab_scene_import_mesh_done);
		}
		else {
			show_box ? project_import_mesh_box(path, true, true, NULL) : import_mesh_run(path, true, true);
		}
		if (drop_x > 0) {
			ui_box_click_to_hide = false; // Prevent closing when going back to window after drag and drop
		}
	}
	else if (path_is_texture(path)) {
		import_texture_run(path, hdr_as_envmap);
		// Place image node
		i32 x0 = ui_nodes_wx;
		i32 x1 = ui_nodes_wx + ui_nodes_ww;
		if (ui_nodes_show && drop_x > x0 && drop_x < x1) {
			i32 asset_index = 0;
			for (i32 i = 0; i < project_assets->length; ++i) {
				if (string_equals(project_assets->buffer[i]->file, path)) {
					asset_index = i;
					break;
				}
			}
			ui_nodes_accept_asset_drop(asset_index);
			ui_nodes_get_nodes()->nodes_drag = false;
			ui_nodes_hwnd->redraws           = 2;
		}

		if (context_raw->tool == TOOL_TYPE_COLORID && project_asset_names->length == 1) {
			ui_header_handle->redraws = 2;
			context_raw->ddirty       = 2;
		}
	}
	else if (path_is_project(path)) {
		import_arm_run_project(path);
	}
	else if (path_is_plugin(path)) {
		import_plugin_run(path);
	}
	else if (path_is_gimp_color_palette(path)) {
		// import_gpl_run(path, false);
	}
	else if (path_is_font(path)) {
		import_font_run(path);
	}
	else if (path_is_folder(path)) {
		import_folder_run(path);
	}
	else {
		if (context_enable_import_plugin(path)) {
			import_asset_run(path, drop_x, drop_y, show_box, true, NULL);
		}
		else {
			console_error(strings_unknown_asset_format());
		}
	}

	if (done != NULL) {
		done();
	}
}
